File: python/ch4-2pointer-1dir/lc406-minimum-size-subarray-sum/using_prefix_sum.py
from typing import (
    List,
)


class Solution:
    """
    @param nums: an array of integers
    @param s: An integer
    @return: an integer representing the minimum size of subarray
    """

    def minimum_size(self, nums: List[int], s: int) -> int:
        # write your code here
        n = len(nums)
        prefix_sum = self.get_prefix_sum(nums)
        min_length = float("inf")
        for start in range(n):
            for end in range(start, n):
                sum_of_subarray = prefix_sum[end + 1] - prefix_sum[start]
                if sum_of_subarray >= s:
                    min_length = min(min_length, end - start + 1)

        if min_length == float("inf"):
            return -1
        return min_length

    def get_prefix_sum(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix_sum = [0]
        for i in range(1, n + 1):
            prefix_sum.append(prefix_sum[i - 1] + nums[i - 1])
        return prefix_sum


solution = Solution()
print(solution.minimum_size([2, 3, 1, 2, 4, 3], 7))


# Don't forget n+1 on line 31 "for i in range(1, n + 1):"
# Dont' forget range on line 19 "for end in range(start, n):"

File: java/src/ch4/lc406_minimum_size_subarray_sum/UsingPrefixSum.java
package ch4.lc406_minimum_size_subarray_sum;

public class UsingPrefixSum {

    public static void main(String[] args) {
        int answer = new UsingPrefixSum().minimumSize(new int[] { 2, 3, 1, 2, 4, 3 }, 7);
        System.out.println("testing result: " + answer);
    }

    public int minimumSize(int[] nums, int s) {
        int minLength = Integer.MAX_VALUE;
        int[] prefixSum = getPrefixSum(nums);

        for (int start = 0; start < nums.length; start++) {
            for (int end = start; end < nums.length; end++) {
                int sumOfSubarray = prefixSum[end + 1] - prefixSum[start];
                if (sumOfSubarray >= s) {
                    minLength = Math.min(minLength, end - start + 1);
                }
            }
        }
        if (minLength == Integer.MAX_VALUE) {
            return -1;
        }
        return minLength;
    }

    private int[] getPrefixSum(int[] nums) {
        int[] prefixSum = new int[nums.length + 1];
        for (int i = 1; i < nums.length + 1; i++) {
            prefixSum[i] = prefixSum[i - 1] + nums[i - 1];
        }
        return prefixSum;
    }

}
